### PR TITLE
Updated Data Model 

### PR DESCRIPTION
**Pull request by suhasHere**
This PR attempts to address #68

the goal of this PR is to
- Define terminologies that is generic (hopefully with less ambiguities) across application domains
- Define relationships between different resources that make up  the data model 
- Specify interactions between entities of the MoQ architecture when dealing with the data model

### CODE REVIEW

**Comment by fluffy** [comment]
This looked like a good generic starting point to me.

File: draft-lcurley-warp.md [hunk 1]
```diff
@@ -175,19 +175,345 @@ x (b):
...
+`
+MediaObjectId :=   WarpStreamId  | GroupId | ObjectId
+`
+
+## Scope / Goals of the Data Model
+
+### Subscribers
+
+One of the goals of the proposed data model is to facilitate subscribers, 
+within a WarpMediSession, to request the right granularity of the media
```
Comment (VMatrix1900) [inline draft-lcurley-warp.md:345]:
```suggestion
within a WarpMediaSession, to request the right granularity of the media 
```

File: draft-lcurley-warp.md [hunk 2]
```diff
@@ -175,19 +175,345 @@ x (b):
...
+given transport mapping.
+
+To that extent, the proposed data model enables following ways for the subscribers 
+(end-points or Relays) to ask for the data:
+
+    -  Request a specific WarpStream. Such a request enables subscribers to 
+       receive everything being published and all future media objects under a 
+       given representation.
+
+    -  Request media to be delivered everything in group currenly being published and
```
Comment (VMatrix1900) [inline draft-lcurley-warp.md:356]:
```suggestion
    -  Request media to be delivered everything in group currently being published and 
```

File: draft-lcurley-warp.md [hunk 3]
```diff
@@ -175,19 +175,345 @@ x (b):
...
+                                                      sub: acme.tv/brodcasts/channel8/alice/sd/group12                                                                          .    
+                                                               ─────.                               
+                                                      ┌──────(  S1   )                              
+                                                      │       `─────'                               
+                                                      │                                             
+            sub: acme.tv/broadcasts/channel8          │   
+┌──────────────┐                ┌──────────────┐      │   
+│              │                │              │      │  sub: acme.tv/brodcasts/channel8/alice/
+|              |                |              |      |       4k        
+│   Ingest     │ ◀──────────────┤  Relay-Edge  │◀─────┘       .─────.
```
Comment (VMatrix1900) [inline draft-lcurley-warp.md:415]: Is Ingest a relay? Is Ingest a publisher side Relay-Edge? By using Relay-edge, are you implying other types of relay such as Relay-Origin/Relay-Central?

File: draft-lcurley-warp.md [hunk 4]
```diff
@@ -175,19 +175,345 @@ x (b):
...
+session or an Interactive conference meeting session or some other applicaiton
+defined session semantics.
+
+`
+Example WarpMediaSession URL for a broadcast 
+  acme.tv/broadcasts/channel8/
+  acme.tv/broadcasts/channel8/
+`
+
+## Media Representation
```
Comment (VMatrix1900) [inline draft-lcurley-warp.md:251]: What is the relationship between the Media Representation and WarpStream? Do they map 1:1? Looks like the Media Representation URI is a verbose ID corresponding to WarpStreamID.

**Comment by vasilvv** [comment]
This PR has two parts; I will go over each part individually.

------

The first part is a rewrite of the existing model section that replaces the broadcast/track/object model with a five-level model.  Since the PR does not diff easily, I am going to attempt to summarize the changes while trying to express my opinion on them one-by-one.

1. The old concept of "broadcast" has been renamed into WarpMediaSession.  I think I like the term "media session" (though I'm not so sure about using CamelCase for all the terms).
1. What was formerly known as "tracks" is now split into "media representation" and "WARP stream".  Those appear to be in one-to-one correspondence, though I do not understand what is the actual difference between those two.  I can't say I am particularly fond of the term "representation", and I object to use of word "stream" for anything that is not a QUIC stream, since that makes things confusing.  If the issue is that "tracks" are too specific to live media, we can use "media feeds", which I think was in the original model PR.
1. A layer of "object groups" is added that reflects the dependencies between objects.  I am not sure an extra level here is useful compared to the previous approach of just annotating the dependency properties directly in the objects themselves, but I do agree that we should have something of this nature.
1. The media objects are still referred to as media objects, though there is some text about encryption and authentication of those.  I don't believe that text belongs in the model, since encryption is fundamentally an issue specific to a particular media format.

------

The second part of this PR is the "Scope / Goals of the Data Model" section that introduces the notion of publishers, subscribers and relays into the draft.  I don't believe this should be a part of the model by itself (it is probably more appropriate to the section describing how the relays work), for two reasons.

The first reason is that this is not applicable to all use cases. For example, it focuses on the notion that the media receivers need to subscribe to things before receiving them, but there are plenty of cases where one side can just open a connection and start sending media (live ingestion, streaming media from a laptop to a smart device, etc). It also suggests a handshake mechanism between the publisher and the origin server, which is not a distinction that has to exist in all scenarios.

The second reason is that I believe that the fact that the media is being relayed should be mostly opaque to the endpoints producing and consuming media.  Setting authentication question aside, a generic HLS player can play media based on a playlist URL alone, not knowing where it's served from (this applies to HTTP in general); similarly, a generic RTMP ingestion client can send media to an RTMP server based on a server URL (plus auth key) alone, without caring what happens next to it.  As player and ingestion client vendors are likely going to constitute the majority of the target audience of this document long-term, I believe the model section should minimize description of parts that are not immediately relevant to them.

--------

My general issue while reviewing this PR is that it tries to address a lot of problems at once, and since it does not explain exactly what are the specific problems it is trying to address, I mostly have to infer those myself.  It would be helpful if the issue #68 specifically spelled out what the issues are; as in, what is the use case affected, what is needed for that specific use case, and why the current model does not allow that to happen.

File: draft-lcurley-warp.md (hunk 3 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:415]: @VMatrix1900 .. Ingest is a publisher in this context and by Relay-Edge , i meant to indicate the relay at the edge. It might be better to term it as just the Relay as you have noticed

File: draft-lcurley-warp.md (hunk 4 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:251]: Media Representation is application level concept and Warp Stream was the protocol level idea where it represent a flow of media objects for a given representation. The separation was to emphasize the fact that each application has their own way to represent such representations but at the MOQ protocol level, they are uniformly called Warp Streams ..

**Comment by suhasHere** [comment]
Responding to  _A layer of "object groups" is added that reflects the dependencies between objects. I am not sure an extra level here is useful compared to the previous approach of just annotating the dependency properties directly in the objects themselves, but I do agree that we should have something of this nature._

Applications group media objects - like WARP does in one group per gop, RUSH does one video frame, some other examples where they group think based on the VR tiles and so on. The dependency relation is application defined and is implicit when an application puts one or more objects in a group.  There is no explicit dependency list carried and is not needed as well.
But having a way to identify group boundaries will help subscribers ask the media at the right granularity and for relays to provide the media objects without having to know about all possible application dependencies. 
The PR has few examples of such grouping and the model is flexible to extend for new class of applications that are evolving in this space.

**Comment by suhasHere** [comment]
Responding to _What was formerly known as "tracks" is now split into "media representation" and "WARP stream". Those appear to be in one-to-one correspondence, though I do not understand what is the actual difference between those two. I can't say I am particularly fond of the term "representation", and I object to use of word "stream" for anything that is not a QUIC stream, since that makes things confusing. If the issue is that "tracks" are too specific to live media, we can use "media feeds", which I think was in the original model PR._

The name tracks is overloaded, representation is somewhere in between (DASH uses it) , Stream for sure is unusable, but not WarpStream. I am not too attached to names, as long as the names are generic to MoQ applications and not just one class of MoQ applications.

**Comment by suhasHere** [comment]
Responding to _The second reason is that I believe that the fact that the media is being relayed should be mostly opaque to the endpoints producing and consuming media. Setting authentication question aside, a generic HLS player can play media based on a playlist URL alone, not knowing where it's served from (this applies to HTTP in general); similarly, a generic RTMP ingestion client can send media to an RTMP server based on a server URL (plus auth key) alone, without caring what happens next to it. As player and ingestion client vendors are likely going to constitute the majority of the target audience of this document long-term, I believe the model section should minimize description of parts that are not immediately relevant to them._

I agree and disagree on the points raised here.  Relays are part of the MoQ architecture. some applications might be totally fine with having no relays but that's not the case for all the applications. Relays might be optional in some architectures and application combinations , in which case the corresponding section of the draft doesn't apply but for the cases where Relays play a role, there is definitely a need for the Relays to understand both the protocol and the data model. This PR addresses that fact.

**Comment by suhasHere** [comment]
Responding to _ As player and ingestion client vendors are likely going to constitute the majority of the target audience of this document long-term, I believe the model section should minimize description of parts that are not immediately relevant to them._

That's not my read of the charter.

File: draft-lcurley-warp.md (hunk 4 above)
Comment (VMatrix1900) [reply draft-lcurley-warp.md:251]: Is the Media Session application level concept or protocol level concept? By data model are we talking about the application level concept or protocol level concept or both?

**Comment by vasilvv** [comment]
> Relays are part of the MoQ architecture. some applications might be totally fine with having no relays but that's not the case for all the applications. Relays might be optional in some architectures and application combinations , in which case the corresponding section of the draft doesn't apply but for the cases where Relays play a role, there is definitely a need for the Relays to understand both the protocol and the data model. 

I think we should be careful to distinguish between the MoQ data model and the MoQ architecture.

Let me explain what I mean here, using HTTP as an example.  HTTP has a fairly simple data model (the protocol exchanges request/response messages that have a method, a target/status code, some header/trailer fields and a body).  It has three wire protocols, all of which are at their heart, just an expression of that data model.  But I don't think there really is such a thing as a single "HTTP architecture".  HTTP is used in an extremely wide range of scenarios, that include direct connections, a wide array of intermediaries, caches, etc.  RFC 9110 acknowledges a lot of them, provides some mechanisms that are only applicable to specific roles (e.g. Section 7.6 describes mechanisms specific to intermediaries), but ultimately, rather than defining a unified "HTTP architecture", it merely describes the common approaches and how they fit together.

Similarly, when talking about MoQ, rather than thinking about a single MoQ architecture, I think our work here mostly revolves around creating the building blocks that one can architect the solution they need.  More specifically, we should:
1. Figure out what are the things being exchanged in the protocol (objects/tracks/broadcasts/sessions/object groups/etc) -- this is the *data model*.
1. Provide some common methods of exchanging those objects (direct push, subscriptions) -- this is also the part of the data model, though there is more flexibility around some parts of this being optional.
1. Enumerate the ways in which we expect those to be used, explaining how you build systems made of boxes that speak MoQ, and what those boxes would do with the data items and methods defined above.

Note that in this scenario, everyone has to understand and implement the first part, but they can only implement the relevant methods and boxes for the architecture they care about.  The only people who have to care about all the possible boxes are people designing the protocol (i.e. us).

(it's 4am on Saturday, so this is me mostly writing my thoughts on this down before I forget -- I'll almost certainly need to elaborate more at the interim)

**TL;DR** "what does MoQ do?" and "what does one do with MoQ?" are two different questions, those should be described in different sections, and relays are mostly an answer to the second question

**Review by kixelated — COMMENTED** [review]
I agree with Victor. It's very difficult to figure out what you want to address and why. Could you try to break it up into smaller issues/PRs for the sake of discussion?

My best guess is:

1. Rename "broadcast" to "WarpMediaSession".
2. Add "Representation", which is member of a broadcast and contains a single track...?
3. Rename "track" to "WarpMediaStream"?
4. Add "WarpObjectsGroup", which is a member of a track and contains objects.
5. Require a globally unique broadcast URL. #70 
6. Add the ability to use an object per slice. #57 
7. Add a sequence number for objects.
8. Add the ability to subscribe starting at a GOP boundary. #52 kinda?
9. Add the ability to subscribe starting at a specific frame?

**Comment by suhasHere** [comment]
> > Relays are part of the MoQ architecture. some applications might be totally fine with having no relays but that's not the case for all the applications. Relays might be optional in some architectures and application combinations , in which case the corresponding section of the draft doesn't apply but for the cases where Relays play a role, there is definitely a need for the Relays to understand both the protocol and the data model.
> 
> I think we should be careful to distinguish between the MoQ data model and the MoQ architecture.
> 
> Let me explain what I mean here, using HTTP as an example. HTTP has a fairly simple data model (the protocol exchanges request/response messages that have a method, a target/status code, some header/trailer fields and a body). It has three wire protocols, all of which are at their heart, just an expression of that data model. But I don't think there really is such a thing as a single "HTTP architecture". HTTP is used in an extremely wide range of scenarios, that include direct connections, a wide array of intermediaries, caches, etc. RFC 9110 acknowledges a lot of them, provides some mechanisms that are only applicable to specific roles (e.g. Section 7.6 describes mechanisms specific to intermediaries), but ultimately, rather than defining a unified "HTTP architecture", it merely describes the common approaches and how they fit together.
> 
> Similarly, when talking about MoQ, rather than thinking about a single MoQ architecture, I think our work here mostly revolves around creating the building blocks that one can architect the solution they need. More specifically, we should:
> 
> 1. Figure out what are the things being exchanged in the protocol (objects/tracks/broadcasts/sessions/object groups/etc) -- this is the _data model_.
> 2. Provide some common methods of exchanging those objects (direct push, subscriptions) -- this is also the part of the data model, though there is more flexibility around some parts of this being optional.
> 3. Enumerate the ways in which we expect those to be used, explaining how you build systems made of boxes that speak MoQ, and what those boxes would do with the data items and methods defined above.
> 
> Note that in this scenario, everyone has to understand and implement the first part, but they can only implement the relevant methods and boxes for the architecture they care about. The only people who have to care about all the possible boxes are people designing the protocol (i.e. us).
> 
> (it's 4am on Saturday, so this is me mostly writing my thoughts on this down before I forget -- I'll almost certainly need to elaborate more at the interim)
> 
> **TL;DR** "what does MoQ do?" and "what does one do with MoQ?" are two different questions, those should be described in different sections, and relays are mostly an answer to the second question

The protocol draft is supposed to talk about protocol elements, data model and how the entities (publishers/subscribers/relays) interact to understand the protocol. If we do it in one section or 3 sections , it is up to us , but the spec should address these things to align with the charter. If we don't thjnk that's the case, may be good thing to raise it to the WG ?

File: draft-lcurley-warp.md [hunk 5]
```diff
@@ -175,19 +175,345 @@ x (b):
...
+
+* Each audio frame is mapped to a WarpObjectGroup. In this grouping, each 
+   WarpObjectGroup has a single audio frame as the WarpMediaObject.
+
+
+## WarpMediaObject/MoqMediaObject/Media Object
+
+Media Objects within Warp belong to a group and carry encoded and 
+encrypted media data. Each object within a group, is identified by 
+sequentially increasing integer, called ObjectId, starting at 0.
```
[the remark this answers is not in the document]
Comment (suhasHere) [reply draft-lcurley-warp.md:327]: > I agree with Victor. It's very difficult to figure out what you want to address and why. Could you try to break it up into smaller issues/PRs for the sake of discussion?
> 
> My best guess is:
> 
> 1. Rename "broadcast" to "WarpMediaSession".
> 2. Add "Representation", which is member of a broadcast and contains a single track...?
> 3. Rename "track" to "WarpMediaStream"?
> 4. Add "WarpObjectsGroup", which is a member of a track, contains objects, and is independently decodable?
> 5. Require a globally unique broadcast URL. #70
> 6. Add the ability to use an object per slice. #57
> 7. Add a sequence number for objects.
> 8. Add the ability to subscribe starting at a GOP boundary. #52 kinda?
> 9. Add the ability to subscribe starting at a specific frame?


We need to work on a data model and it needs to be dealt in one place. Splitting across the PRs will just cause more confusions and it's hard to have discussions made coherently.

File: draft-lcurley-warp.md [hunk 6]
```diff
@@ -175,19 +175,345 @@ x (b):
...
+to audio stream from alice, whereas S2 subscribers
+to all the streams being published. Relay at the Edge
+forwards the unique subscriptions to the downstream Realys,
+as needed, to setup the delivery network.
+
+
+### Publishers
+
+Publishers generate media objects with in a WarpMediaSession. A given
+publishers start by obtaining `WarpStreamId` via out-of-band
```
[the remark this answers is not in the document]
Comment (suhasHere) [reply draft-lcurley-warp.md:488]: There is no negotiation here. can you provide details on the flow of steps that happens today with these generic clients...

File: draft-lcurley-warp.md [hunk 7]
```diff
@@ -175,19 +175,345 @@ x (b):
...
+wherever applicable.
+
+A WarpMediaSession can represent a Broadcast for a live media 
+session or an Interactive conference meeting session or some other applicaiton
+defined session semantics.
+
+`
+Example WarpMediaSession URL for a broadcast 
+  acme.tv/broadcasts/channel8/
+  acme.tv/broadcasts/channel8/
```
[the remark this answers is not in the document]
Comment (suhasHere) [reply draft-lcurley-warp.md:248]: That should have been 2 different examples. Copy paste error

File: draft-lcurley-warp.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (kixelated) [reply draft-lcurley-warp.md:488]: I guess today publishers will get a URL. This would require a URL (for WebTransport) and another URL for the `WarpStreamId`. Actually aren't there multiple `WarpStreamIds`? Do you need to pre-negotiate one for every rendition?

File: draft-lcurley-warp.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kixelated) [reply draft-lcurley-warp.md:327]: I want to discuss each of these separately. 

A few of them are good changes (6, 7, 8). Some of the changes are up for debate and deserve their own issue, like (1, 4, 5, 9). Some of the changes are just questionable (2, 3) and really just drag the rest of the PR down. 

But you absolutely don't need to rewrite the object model. It's perfectly acceptable to create a PR that simultaneously amends the object model while adding the corresponding wire format. I really don't think there's a huge discrepancy between the current object model and the one that you're proposing.

**Comment by suhasHere** [comment]
Responding to _As player and ingestion client vendors are likely going to constitute the majority of the target audience of this document long-term, I believe the model section should minimize description of parts that are not immediately relevant to them_

This might be a good thing to bring to the WG and the list ..

File: draft-lcurley-warp.md (hunk 4 above)
Comment (suhasHere) [reply draft-lcurley-warp.md:251]: Here is what I had in mind, do let me know if it helps or should we adjust it differently.

Each MoQ application will have their own way to identify publisher media capabilities that list media types, their quality attributes (codec, bitrates and so on). They are typically some sort of textual descriptions. Then for each of those there will be stream of grouped media objects that is actually sent.

I totally understand the reason why this might cause confusion. It might be better served, it we can just say 

```
Warp Media Streams identify media objects belonging to a publisher's media capability/representation. A given Media Representation, in an application defined way,  identifies capabilities of a publisher's media source(s) within a WarpMediaSession. Representations identifies things such as media source type, codec info, bitrate and other attribute qualities.

```

Would this work ?

Comment (kixelated) [reply draft-lcurley-warp.md:251]: Yeah that won't work. We need to support generic clients. #71

Any information required to broadcast or view needs to be part of the protocol. That can be as simple as specifying container/manifest formats with enough information, but it cannot be left up to the application.

File: draft-lcurley-warp.md [hunk 8]
```diff
@@ -175,19 +175,329 @@ x (b):
...
+       receive all future MediaGroups and MediaObjects under a 
+       given MediaStream.
+    -  Request a specific MediaStream starting at a specific MediaGroup. Since the MediaGroup 
+       identification is ordinal, subscribers may request for example content starting from 
+       the last, or 3rd last, MediaGroup available in the cache, or from the first one for 
+       which MediaGroupID  > N
+    - Request a specific MediaSession. Such a request enables subscribers to 
+       receive all future MediaStreams, MediaGroups and MediaObjects under a 
+       given MediaSession.
+
```
Comment (afrind) [inline draft-lcurley-warp.md:349]: This text appears duplicated with 336-338

File: draft-lcurley-warp.md [hunk 9]
```diff
@@ -175,19 +175,329 @@ x (b):
...
+       identification is ordinal, subscribers may request for example content starting from 
+       the last, or 3rd last, MediaGroup available in the cache, or from the first one for 
+       which MediaGroupID  > N
+    - Request a specific MediaSession. Such a request enables subscribers to 
+       receive all future MediaStreams, MediaGroups and MediaObjects under a 
+       given MediaSession.
+       
+The MediSession ID to given to each subscriber via an out-of-band mechanism. The subscriber uses this 
+MediaSessionID to request the CATALOG, which in turn provides information to the client about the 
+MediaStreamIDs which it can use to construct its subscription requests.
```
Comment (afrind) [inline draft-lcurley-warp.md:352]: Why out of band?

Also, better to use publisher/consumer rather than client/server terminology, since for ingest, the client is the publisher.

File: draft-lcurley-warp.md [hunk 10]
```diff
@@ -175,19 +175,329 @@ x (b):
...
+pub: acme.tv/broadcasts/channel8/alice/hd                    `─────'                              
+pub: acme.tv/broadcasts/channel8/alice/sd                                                                   
+pub: acme.tv/broadcasts/channel8/alice/4k            sub: acme.tv/brodcasts/channel8/alice/sd  
+                                                     sub: acme.tv/brodcasts/channel8/alice/hd   
+
+~~~~                                                                                                                             
+  
+The relay does not intercept and parse the CATALOG messages, therefore it does not know the entireity of 
+the content being produced by Alice. It simply aggregates and forwards all subscription requests that 
+it receives.
```
Comment (afrind) [inline draft-lcurley-warp.md:420]: I think there are features/performance that can be unlocked by a "media gateway" that does parse the CATALOG, such as the ability for a subscriber to ask the gateway to automatically switch between qualities within a latency bound, but perhaps such a function doesn't need to be considered a "relay".

**Comment by afrind** [comment]
What's the disposition of this PR in light of the newer one?  Should we close it, or do we intend to iterate on both?

**Comment by afrind** [comment]
Has this been overtaken by events, such as #121?  Can we close?
